### PR TITLE
Device By Path via Refactor

### LINKED ISSUE #38: How to handle multiple connected devices?
I've found that when I have multiple devices connected, the library sometimes selects the wrong one to use. I would like it if, given a key handle, the library would determine automatically what device can fulfill the request but my experience has been that it throws an error if it happens to choose the wrong device.

Is this a mistake I'm making in the way I'm configuring my usage of this code?

### PR DESCRIPTION
**Pull request by obelisk**
This is the refactor of the multiple device support PR with the refactor of some of the commands into a submodule of FidoKey.

Closes #38

### CODE REVIEW

**Comment by obelisk** [comment]
Tests are here: https://github.com/obelisk/ctap-hid-fido2/pull/3

**Comment by obelisk** [comment]
Done refactoring things under `FidoKeyHid`! This is how I think the API should work because it means once you get a device you can be sure you're contacting the correct device syntactically.

`device.make_credential` I think is better than `ctap_hid_fido2::make_credential(HidParam::Path("device_path"))`.

All modules have been refactored to this new convention but I still need to update the examples and test everything.

However! This is what it would look like :)

**Comment by obelisk** [comment]
Put up a new version of the Wink example. I was going to do all of them but then realized that I'll have mega merge conflicts if I don't rebase on #49 properly

**Comment by obelisk** [comment]
Closes #38

**Comment by obelisk** [comment]
How are you feeling about this restructuring? I can rebase this, and fix the merge conflicts, but looking at `wink`, what do you think of the new design?

I think it make the code a lot more organized and easier to use. It also makes `lib.rs` much easier to read and understand.

**Comment by gebogebogebo** [comment]
@obelisk 

Sorry, I have not had time to check it because I have been busy with my business.
I'm checking the Develop branch now. 
I will release 2.2.4 when we are done.
Then check this PR.
This PR cannot be released as ver2 because it contains destructive changes.

Please give some more time.I need about a week for code review.

**Comment by gebogebogebo** [comment]
I'd like to run it and see.
Can you please make the build pass with minimal modifications?

File: src/hid/mod.rs [hunk 1]
```diff
@@ -20,9 +19,57 @@ pub struct DeviceInfo {
     pub report_size: u16,
 }
 
+/// HID device vendor ID , product ID
+#[derive(Clone)]
+pub enum HidParam {
+    /// Specified when looking for any FIDO device of a certain kind
+    VidPid {vid: u16, pid: u16},
```
Comment (gebogebogebo) [inline src/hid/mod.rs:26]: With the addition of VidPid, LibCfg.hid_params will likely be unnecessary.

**Comment by obelisk** [comment]
Will try to fix this today :)

**Comment by obelisk** [comment]
Going to take me a while to port all the examples so you can test them. So far `wink` and `test-with-pin-rk` are done. Will keep pushing commits as I finish up each example.

I've also rebased on master so it should have the new HMAC changes as well.

**Comment by obelisk** [comment]
@gebogebogebo All examples ported and compiling! I've tested a few of them but the more complex ctapcli I haven't had a chance yet.

I think everything should work (if I did my job right) as this is just a refactor :)

**Comment by gebogebogebo** [comment]
@obelisk 
Many thanks for all the improvements!
Multi-device support would be great!
Can I merge this PR now?

**Comment by obelisk** [comment]
Did it pass all your tests with all your different devices? This is a pretty sprawling change that really changes how the library is used.

There is also the possibility that some members will need to be visibility adjusted (like `Extension` in make_credential and get_assertion).

If you've read through the code, and it passes your tests, then I think it's good to merge but we should definitely let it soak in develop for a while so we can make sure it's solid.

**Comment by obelisk** [comment]
Oh there is actually an issue, the cargo version isn't set correctly

Let me fix that

**Comment by gebogebogebo** [comment]
First, merge it into the develop branch and test it with the other modifications.
When testing is complete, merge into the master branch.
It may take a little time to merge into the master branch.
Still, I think this improvement is very worthwhile!

**Pull request merged**